FEXCore/include/FEXCore/Utils/ThreadPoolAllocator.h

```diff
@@ -120,16 +120,16 @@ class IntrusivePooledAllocator {
   }
 
   /**
-   * @brief Immediately release the buffer back to the allocator
+   * @brief Immediately release the buffer back to the allocator, given it has not been reclaimed
    *
    * @param Buffer - The iterator that was previously given with ClaimBuffer
-   *
-   * Once this is called on a buffer then the pool allocator has full ownership of the buffer
    */
-  void UnclaimBuffer(ContainerType::iterator Buffer) {
-    std::unique_lock lk {AllocationMutex};
-    (*Buffer)->CurrentClientOwnedFlag->store(ClientFlags::FLAG_FREE);
-    UnclaimBufferImpl(Buffer);
+  void UnclaimBuffer(ContainerType::iterator Buffer, BufferOwnedFlag* ClientFlag) {
+    // Transition the buffer to free, unclaiming if it wasn't free prior.
+    if (ClientFlag->exchange(ClientFlags::FLAG_FREE) != ClientFlags::FLAG_FREE) {
+      std::unique_lock lk {AllocationMutex};
```

Comment:
Can't this now still end up in the situation that the flag goes from {DISOWNED,OWNED} to FREE, try and acquire the lock, failing, another thread scoops up the buffer, and then this executes freeing the buffer that another thread now owns?

I think the previous implementation where it checks the owned status, acquires the lock, and then exchanges the owned status is the most bug free.

Reply:
Perhaps I'm missing something, but it shouldn't be possible for another thread to claim the buffer until it's in UnclaimedBuffers, and that only happens after the lock is acquired.

Reply:
```
T0.UnclaimBuffer(T0Buffer, Flag) -> T0 unclaiming buffer
T0.exchange(Flag, Free) == Owned; -> Free's buffer

=== This can weasel its way in here ===
T1.ClaimBuffer() returns T0Buffer -> Claims buffer that T0 just freed.

T0.Lock(AllocationMutex) -> Locks the allocation mutex, only gains ownership when T1.ClaimBuffer() is done
T0.UnclaimBuffer(T0Buffer) -> T0 free's the buffer that T1 just claimed, assuming it still owns it.
```

Reply:
It's a little tricky since `ClaimBuffer` walks both claimed buffers and freed buffer lists looking for disowned buffers inside of the ClaimedBuffers array, and `UnclaimedBuffers` is always free buffers.

Reply:
T1.ClaimBuffer() could never return the buffer though, as it isn't DISOWNED and in the ClaimedBuffers list, nor is it in the UnclaimedBuffers list

Reply:
Ah, you are probably right, safe since it only checks for disowned, not free.

Reply:
The joys of lock free code :)